### ISSUE TITLE
lib/warnings.t fails on Alpine Linux

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#16835]
Migrated from [rt.perl.org#133819](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=133819) (status was 'resolved')

Searchable as RT133819$

**Comment by p5pRT** [comment]
### From @tonycoz

#### Created by @tonycoz



alpine64&#8203;:\~/dev/perl/git/perl/t$ \./perl harness \.\./lib/warnings\.t
\.\./lib/warnings\.t \.\. 704/902 PROG&#8203;: 
require '\.\./loc\_tools\.pl';
unless \(locales\_enabled\(\)\) \{
&nbsp;    print\("SKIPPED\\n\# locales not available\\n"\)\,exit;
\}
eval \{ require POSIX; POSIX\->import\("locale\_h"\) \};
if \($@&#8203;\) \{
&nbsp;    print\("SKIPPED\\n\# no POSIX\\n"\)\,exit;
\}
my $utf8\_locale = find\_utf8\_ctype\_locale\(\);
unless \($utf8\_locale\) \{
&nbsp;    print\("SKIPPED\\n\# No UTF\-8 locale available\\n"\)\,exit;
\}
use warnings 'locale';
use locale;
setlocale\(&POSIX&#8203;::LC\_CTYPE\, "C"\);
 "abc def" =~ /\\b\{wb\}\.\*?/;
 "abc def" =~ /\\B\{wb\}\.\*?/;
setlocale\(&POSIX&#8203;::LC\_CTYPE\, $utf8\_locale\);
 "abc def" =~ /\\b\{wb\}\.\*?/;
 "abc def" =~ /\\B\{wb\}\.\*?/;
EXPECTED&#8203;:
Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8 locale at \- line 16\.
Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8 locale at \- line 16\.
Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8 locale at \- line 17\.
Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8 locale at \- line 17\.
Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8 locale at \- line 17\.
GOT&#8203;:
Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8 locale at \- line 16\, \<DATA> line 47\.
Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8 locale at \- line 16\, \<DATA> line 47\.
Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8 locale at \- line 17\, \<DATA> line 47\.
Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8 locale at \- line 17\, \<DATA> line 47\.
Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8 locale at \- line 17\, \<DATA> line 47\.
\# Failed test 704 \- \\b\{\} in UTF\-8 locale at lib/warnings/regexec line 189
\.\./lib/warnings\.t \.\. Failed 1/902 subtests 
&nbsp;    \(less 3 skipped subtests&#8203;: 898 okay\)

Test Summary Report
\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
\.\./lib/warnings\.t \(Wstat&#8203;: 0 Tests&#8203;: 902 Failed&#8203;: 1\)
&nbsp;  Failed test&#8203;:    704
Files=1\, Tests=902\,  7 wallclock secs \( 0\.08 usr  0\.00 sys \+  6\.50 cusr  0\.74 csys =  7\.32 CPU\)
Result&#8203;: FAIL

I don't know why we're getting the \<DATA> references in the output
on Alpine\, but not on other platforms\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.29.8:

Configured by tony at Mon Feb  4 13:22:49 AEDT 2019.

Summary of my perl5 (revision 5 version 29 subversion 8) configuration:
  Commit id: 04db542212fdad3a62f13afe741c99028f4bf799
  Platform:
    osname=linux
    osvers=4.19.18-0-vanilla
    archname=x86_64-linux
    uname='linux alpine64 4.19.18-0-vanilla #1-alpine smp sat jan 26 21:59:33 utc 2019 x86_64 linux '
    config_args='-des -Dusedevel'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=undef
    usemultiplicity=undef
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
    bincompat5005=undef
  Compiler:
    cc='cc'
    ccflags ='-D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    optimize='-O2'
    cppflags='-D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong'
    ccversion=''
    gccversion='8.2.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='cc'
    ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/include/fortify /usr/lib /usr/local/lib /lib/../lib /usr/lib/../lib /lib
    libs=-lpthread -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -ldl -lm -lcrypt -lutil -lc
    libc=/usr/lib/libc.a
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E'
    cccdlflags='-fPIC'
    lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector-strong'



@INC for perl 5.29.8:
    lib
    /usr/local/lib/perl5/site_perl/5.29.8/x86_64-linux
    /usr/local/lib/perl5/site_perl/5.29.8
    /usr/local/lib/perl5/5.29.8/x86_64-linux
    /usr/local/lib/perl5/5.29.8


Environment for perl 5.29.8:
    HOME=/home/tony
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
    PERL_BADLANG (unset)
    SHELL=/bin/ash

```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Mon\, 04 Feb 2019 03&#8203;:28&#8203;:07 GMT\, tonyc wrote&#8203;:
> This is a bug report for perl from tony@&#8203;develop\-help\.com\,
> generated with the help of perlbug 1\.41 running under perl 5\.29\.8\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> alpine64&#8203;:\~/dev/perl/git/perl/t$ \./perl harness \.\./lib/warnings\.t
>  \.\./lib/warnings\.t \.\. 704/902 PROG&#8203;:
> require '\.\./loc\_tools\.pl';
> unless \(locales\_enabled\(\)\) \{
>     print\("SKIPPED\\n\# locales not available\\n"\)\,exit;
> \}
> eval \{ require POSIX; POSIX\->import\("locale\_h"\) \};
> if \($@&#8203;\) \{
>     print\("SKIPPED\\n\# no POSIX\\n"\)\,exit;
> \}
> my $utf8\_locale = find\_utf8\_ctype\_locale\(\);
> unless \($utf8\_locale\) \{
>     print\("SKIPPED\\n\# No UTF\-8 locale available\\n"\)\,exit;
> \}
> use warnings 'locale';
> use locale;
> setlocale\(&POSIX&#8203;::LC\_CTYPE\, "C"\);
>  "abc def" =~ /\\b\{wb\}\.\*?/;
>  "abc def" =~ /\\B\{wb\}\.\*?/;
> setlocale\(&POSIX&#8203;::LC\_CTYPE\, $utf8\_locale\);
>  "abc def" =~ /\\b\{wb\}\.\*?/;
>  "abc def" =~ /\\B\{wb\}\.\*?/;
> EXPECTED&#8203;:
> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
> locale at \- line 16\.
> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
> locale at \- line 16\.
> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
> locale at \- line 17\.
> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
> locale at \- line 17\.
> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
> locale at \- line 17\.
> GOT&#8203;:
> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
> locale at \- line 16\, \<DATA> line 47\.
> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
> locale at \- line 16\, \<DATA> line 47\.
> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
> locale at \- line 17\, \<DATA> line 47\.
> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
> locale at \- line 17\, \<DATA> line 47\.
> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
> locale at \- line 17\, \<DATA> line 47\.
> \# Failed test 704 \- \\b\{\} in UTF\-8 locale at lib/warnings/regexec line
> 189
>  \.\./lib/warnings\.t \.\. Failed 1/902 subtests
>         \(less 3 skipped subtests&#8203;: 898 okay\)
> 
> Test Summary Report
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \.\./lib/warnings\.t \(Wstat&#8203;: 0 Tests&#8203;: 902 Failed&#8203;: 1\)
>   Failed test&#8203;:  704
> Files=1\, Tests=902\,  7 wallclock secs \( 0\.08 usr  0\.00 sys \+  6\.50
> cusr  0\.74 csys =  7\.32 CPU\)
> Result&#8203;: FAIL
> 
> I don't know why we're getting the \<DATA> references in the output
> on Alpine\, but not on other platforms\.
> 

Can you see whether this behavior bisects to this commit?

\#\#\#\#\#
commit 8c165a32b7cc4f2a147a37c920a96f1b09b2386d
Author&#8203;:     Karl Williamson \<khw@&#8203;cpan\.org>
AuthorDate&#8203;: Sun Dec 30 21&#8203;:29&#8203;:07 2018 \-0700
Commit&#8203;:     Karl Williamson \<khw@&#8203;cpan\.org>
CommitDate&#8203;: Sat Jan 5 10&#8203;:42&#8203;:26 2019 \-0700

regen/warnings\.pl&#8203;: Fix undefined C behavior

This fixes compiler warnings "performing pointer arithmetic on a null pointer has undefined behavior"

There are several ways to fix this\.  This one was suggested by Tomasz Konojacki\+\+\.  Instead of trying to point to address 1 and 2\, two  variables are created\, and we point to them\.  const is cast away\.
\#\#\#\#\#

Thank you very much\.
Jim Keenan




> \[Please do not change anything below this line\]
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \-\-\-
> Flags&#8203;:
>     category=core
>     severity=low
> \-\-\-
> Site configuration information for perl 5\.29\.8&#8203;:
> 
> Configured by tony at Mon Feb  4 13&#8203;:22&#8203;:49 AEDT 2019\.
> 
> Summary of my perl5 \(revision 5 version 29 subversion 8\)
> configuration&#8203;:
>   Commit id&#8203;: 04db542212fdad3a62f13afe741c99028f4bf799
>   Platform&#8203;:
>     osname=linux
>     osvers=4\.19\.18\-0\-vanilla
>     archname=x86\_64\-linux
>     uname='linux alpine64 4\.19\.18\-0\-vanilla \#1\-alpine smp sat jan 26
> 21&#8203;:59&#8203;:33 utc 2019 x86\_64 linux '
>     config\_args='\-des \-Dusedevel'
>     hint=recommended
>     useposix=true
>     d\_sigaction=define
>     useithreads=undef
>     usemultiplicity=undef
>     use64bitint=define
>     use64bitall=define
>     uselongdouble=undef
>     usemymalloc=n
>     default\_inc\_excludes\_dot=define
>     bincompat5005=undef
>   Compiler&#8203;:
>     cc='cc'
>     ccflags ='\-D\_GNU\_SOURCE \-fwrapv \-fno\-strict\-aliasing \-pipe
> \-fstack\-protector\-strong \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'
>     optimize='\-O2'
>     cppflags='\-D\_GNU\_SOURCE \-fwrapv \-fno\-strict\-aliasing \-pipe
> \-fstack\-protector\-strong'
>     ccversion=''
>     gccversion='8\.2\.0'
>     gccosandvers=''
>     intsize=4
>     longsize=8
>     ptrsize=8
>     doublesize=8
>     byteorder=12345678
>     doublekind=3
>     d\_longlong=define
>     longlongsize=8
>     d\_longdbl=define
>     longdblsize=16
>     longdblkind=3
>     ivtype='long'
>     ivsize=8
>     nvtype='double'
>     nvsize=8
>     Off\_t='off\_t'
>     lseeksize=8
>     alignbytes=8
>     prototype=define
>   Linker and Libraries&#8203;:
>     ld='cc'
>     ldflags =' \-fstack\-protector\-strong \-L/usr/local/lib'
>     libpth=/usr/include/fortify /usr/lib /usr/local/lib /lib/\.\./lib
> /usr/lib/\.\./lib /lib
>     libs=\-lpthread \-ldl \-lm \-lcrypt \-lutil \-lc
>     perllibs=\-lpthread \-ldl \-lm \-lcrypt \-lutil \-lc
>     libc=/usr/lib/libc\.a
>     so=so
>     useshrplib=false
>     libperl=libperl\.a
>     gnulibc\_version=''
>   Dynamic Linking&#8203;:
>     dlsrc=dl\_dlopen\.xs
>     dlext=so
>     d\_dlsymun=undef
>     ccdlflags='\-Wl\,\-E'
>     cccdlflags='\-fPIC'
>     lddlflags='\-shared \-O2 \-L/usr/local/lib \-fstack\-protector\-strong'
> 
> 
> \-\-\-
> @&#8203;INC for perl 5\.29\.8&#8203;:
>     lib
>     /usr/local/lib/perl5/site\_perl/5\.29\.8/x86\_64\-linux
>     /usr/local/lib/perl5/site\_perl/5\.29\.8
>     /usr/local/lib/perl5/5\.29\.8/x86\_64\-linux
>     /usr/local/lib/perl5/5\.29\.8
> 
> \-\-\-
> Environment for perl 5\.29\.8&#8203;:
>     HOME=/home/tony
>     LANG \(unset\)
>     LANGUAGE \(unset\)
>     LD\_LIBRARY\_PATH \(unset\)
>     LOGDIR \(unset\)
>     PATH=/usr/local/sbin&#8203;:/usr/local/bin&#8203;:/usr/sbin&#8203;:/usr/bin&#8203;:/sbin&#8203;:/bin
>     PERL\_BADLANG \(unset\)
>     SHELL=/bin/ash


\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @khwilliamson

On 2/4/19 6&#8203;:46 AM\, James E Keenan via RT wrote&#8203;:
> On Mon\, 04 Feb 2019 03&#8203;:28&#8203;:07 GMT\, tonyc wrote&#8203;:
>> This is a bug report for perl from tony@&#8203;develop\-help\.com\,
>> generated with the help of perlbug 1\.41 running under perl 5\.29\.8\.
>>
>>
>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>> \[Please describe your issue here\]
>>
>> alpine64&#8203;:\~/dev/perl/git/perl/t$ \./perl harness \.\./lib/warnings\.t
>>   \.\./lib/warnings\.t \.\. 704/902 PROG&#8203;:
>> require '\.\./loc\_tools\.pl';
>> unless \(locales\_enabled\(\)\) \{
>>      print\("SKIPPED\\n\# locales not available\\n"\)\,exit;
>> \}
>> eval \{ require POSIX; POSIX\->import\("locale\_h"\) \};
>> if \($@&#8203;\) \{
>>      print\("SKIPPED\\n\# no POSIX\\n"\)\,exit;
>> \}
>> my $utf8\_locale = find\_utf8\_ctype\_locale\(\);
>> unless \($utf8\_locale\) \{
>>      print\("SKIPPED\\n\# No UTF\-8 locale available\\n"\)\,exit;
>> \}
>> use warnings 'locale';
>> use locale;
>> setlocale\(&POSIX&#8203;::LC\_CTYPE\, "C"\);
>>   "abc def" =~ /\\b\{wb\}\.\*?/;
>>   "abc def" =~ /\\B\{wb\}\.\*?/;
>> setlocale\(&POSIX&#8203;::LC\_CTYPE\, $utf8\_locale\);
>>   "abc def" =~ /\\b\{wb\}\.\*?/;
>>   "abc def" =~ /\\B\{wb\}\.\*?/;
>> EXPECTED&#8203;:
>> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
>> locale at \- line 16\.
>> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
>> locale at \- line 16\.
>> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
>> locale at \- line 17\.
>> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
>> locale at \- line 17\.
>> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
>> locale at \- line 17\.
>> GOT&#8203;:
>> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
>> locale at \- line 16\, \<DATA> line 47\.
>> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
>> locale at \- line 16\, \<DATA> line 47\.
>> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
>> locale at \- line 17\, \<DATA> line 47\.
>> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
>> locale at \- line 17\, \<DATA> line 47\.
>> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
>> locale at \- line 17\, \<DATA> line 47\.
>> \# Failed test 704 \- \\b\{\} in UTF\-8 locale at lib/warnings/regexec line
>> 189
>>   \.\./lib/warnings\.t \.\. Failed 1/902 subtests
>>          \(less 3 skipped subtests&#8203;: 898 okay\)
>>
>> Test Summary Report
>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>> \.\./lib/warnings\.t \(Wstat&#8203;: 0 Tests&#8203;: 902 Failed&#8203;: 1\)
>>    Failed test&#8203;:  704
>> Files=1\, Tests=902\,  7 wallclock secs \( 0\.08 usr  0\.00 sys \+  6\.50
>> cusr  0\.74 csys =  7\.32 CPU\)
>> Result&#8203;: FAIL
>>
>> I don't know why we're getting the \<DATA> references in the output
>> on Alpine\, but not on other platforms\.
>>
> 
> Can you see whether this behavior bisects to this commit?
> 
> \#\#\#\#\#
> commit 8c165a32b7cc4f2a147a37c920a96f1b09b2386d
> Author&#8203;:     Karl Williamson \<khw@&#8203;cpan\.org>
> AuthorDate&#8203;: Sun Dec 30 21&#8203;:29&#8203;:07 2018 \-0700
> Commit&#8203;:     Karl Williamson \<khw@&#8203;cpan\.org>
> CommitDate&#8203;: Sat Jan 5 10&#8203;:42&#8203;:26 2019 \-0700
> 
> regen/warnings\.pl&#8203;: Fix undefined C behavior
> 
> This fixes compiler warnings "performing pointer arithmetic on a null pointer has undefined behavior"
> 
> There are several ways to fix this\.  This one was suggested by Tomasz Konojacki\+\+\.  Instead of trying to point to address 1 and 2\, two  variables are created\, and we point to them\.  const is cast away\.
> \#\#\#\#\#
> 
> Thank you very much\.
> Jim Keenan
> 
> 

No need\.  Tony and I figured this out on irc\, and some later snooping
I'll explain it in the commit message\.

**Comment by p5pRT** [comment]
### From @tonycoz

On Sun\, 03 Feb 2019 19&#8203;:28&#8203;:07 \-0800\, tonyc wrote&#8203;:
> alpine64&#8203;:\~/dev/perl/git/perl/t$ \./perl harness \.\./lib/warnings\.t
>  \.\./lib/warnings\.t \.\. 704/902 PROG&#8203;:
> require '\.\./loc\_tools\.pl';
> unless \(locales\_enabled\(\)\) \{
>     print\("SKIPPED\\n\# locales not available\\n"\)\,exit;
> \}
> eval \{ require POSIX; POSIX\->import\("locale\_h"\) \};
> if \($@&#8203;\) \{
>     print\("SKIPPED\\n\# no POSIX\\n"\)\,exit;
> \}
> my $utf8\_locale = find\_utf8\_ctype\_locale\(\);
> unless \($utf8\_locale\) \{
>     print\("SKIPPED\\n\# No UTF\-8 locale available\\n"\)\,exit;
> \}
> use warnings 'locale';
> use locale;
> setlocale\(&POSIX&#8203;::LC\_CTYPE\, "C"\);
>  "abc def" =~ /\\b\{wb\}\.\*?/;
>  "abc def" =~ /\\B\{wb\}\.\*?/;
> setlocale\(&POSIX&#8203;::LC\_CTYPE\, $utf8\_locale\);
>  "abc def" =~ /\\b\{wb\}\.\*?/;
>  "abc def" =~ /\\B\{wb\}\.\*?/;
> EXPECTED&#8203;:
> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
> locale at \- line 16\.
> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
> locale at \- line 16\.
> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
> locale at \- line 17\.
> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
> locale at \- line 17\.
> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
> locale at \- line 17\.
> GOT&#8203;:
> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
> locale at \- line 16\, \<DATA> line 47\.
> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
> locale at \- line 16\, \<DATA> line 47\.
> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
> locale at \- line 17\, \<DATA> line 47\.
> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
> locale at \- line 17\, \<DATA> line 47\.
> Use of \\b\{\} or \\B\{\} for non\-UTF\-8 locale is wrong\.  Assuming a UTF\-8
> locale at \- line 17\, \<DATA> line 47\.
> \# Failed test 704 \- \\b\{\} in UTF\-8 locale at lib/warnings/regexec line
> 189
>  \.\./lib/warnings\.t \.\. Failed 1/902 subtests
>         \(less 3 skipped subtests&#8203;: 898 okay\)
> 
> Test Summary Report
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \.\./lib/warnings\.t \(Wstat&#8203;: 0 Tests&#8203;: 902 Failed&#8203;: 1\)
>   Failed test&#8203;:  704
> Files=1\, Tests=902\,  7 wallclock secs \( 0\.08 usr  0\.00 sys \+  6\.50
> cusr  0\.74 csys =  7\.32 CPU\)
> Result&#8203;: FAIL
> 
> I don't know why we're getting the \<DATA> references in the output
> on Alpine\, but not on other platforms\.

This has been fixed in blead\.

Tony

**Comment by p5pRT** [comment]
@tonycoz - Status changed from 'open' to 'pending release'

**Comment by p5pRT** [comment]
### From @khwilliamson

Thank you for filing this report\.  You have helped make Perl better\.

With the release today of Perl 5\.30\.0\, this and 160 other issues have been
resolved\.

Perl 5\.30\.0 may be downloaded via&#8203;:
https://metacpan.org/release/XSAWYERX/perl-5.30.0

If you find that the problem persists\, feel free to reopen this ticket\.

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'pending release' to 'resolved'

**Comment by MatthewZMD** [comment]
The same error appeared again on a poky build from the Yocto Project, built using `bitbake core-image-sato`

This is discovered while tackling a [bug](https://bugzilla.yoctoproject.org/show_bug.cgi?id=13296) on Yocto.

To reproduce it, please refer to [Yocto Project Quick Build](https://www.yoctoproject.org/docs/2.5/brief-yoctoprojectqs/brief-yoctoprojectqs.html) first for anyone unfamiliar with the Yocto project.

1. Initialize build environment as per the Quick Build. Append following to `.../build/conf/local.conf` file:
```
DISTRO_FEATURES_append = " ptest"
CORE_IMAGE_EXTRA_INSTALL += " perl-ptest"
```
2. Run `bitbake core-image-sato` and `runqemu qemux86-64 kvm` to boot the image.
3. Login with root without any password
4.
```
# cd /usr/lib/perl/ptest/t/
# ./perl harness ../lib/warnings.t 
../lib/warnings.t .. 704/903 PROG: 
require '../loc_tools.pl';
unless (locales_enabled()) {
    print("SKIPPED\n# locales not available\n"),exit;
}
eval { require POSIX; POSIX->import("locale_h") };
if ($@) {
    print("SKIPPED\n# no POSIX\n"),exit;
}
my $utf8_locale = find_utf8_ctype_locale();
unless ($utf8_locale) {
    print("SKIPPED\n# No UTF-8 locale available\n"),exit;
}
use warnings 'locale';
use locale;
setlocale(&POSIX::LC_CTYPE, "C");
 "abc def" =~ /\b{wb}.*?/;
 "abc def" =~ /\B{wb}.*?/;
setlocale(&POSIX::LC_CTYPE, $utf8_locale);
 "abc def" =~ /\b{wb}.*?/;
 "abc def" =~ /\B{wb}.*?/;
EXPECTED:
Use of \b{} or \B{} for non-UTF-8 locale is wrong.  Assuming a UTF-8 locale at - line 16.
Use of \b{} or \B{} for non-UTF-8 locale is wrong.  Assuming a UTF-8 locale at - line 16.
Use of \b{} or \B{} for non-UTF-8 locale is wrong.  Assuming a UTF-8 locale at - line 17.
Use of \b{} or \B{} for non-UTF-8 locale is wrong.  Assuming a UTF-8 locale at - line 17.
Use of \b{} or \B{} for non-UTF-8 locale is wrong.  Assuming a UTF-8 locale at - line 17.
GOT:
Use of \b{} or \B{} for non-UTF-8 locale is wrong.  Assuming a UTF-8 locale at - line 16, <DATA> line 47.
Use of \b{} or \B{} for non-UTF-8 locale is wrong.  Assuming a UTF-8 locale at - line 16, <DATA> line 47.
Use of \b{} or \B{} for non-UTF-8 locale is wrong.  Assuming a UTF-8 locale at - line 17, <DATA> line 47.
Use of \b{} or \B{} for non-UTF-8 locale is wrong.  Assuming a UTF-8 locale at - line 17, <DATA> line 47.
Use of \b{} or \B{} for non-UTF-8 locale is wrong.  Assuming a UTF-8 locale at - line 17, <DATA> line 47.
# Failed test 704 - \b{} in UTF-8 locale at lib/warnings/regexec line 189
../lib/warnings.t .. Failed 1/903 subtests 
        (less 3 skipped subtests: 899 okay)

Test Summary Report
-------------------
../lib/warnings.t (Wstat: 0 Tests: 903 Failed: 1)
  Failed test:  704
Files=1, Tests=903, 15 wallclock secs ( 0.09 usr  0.01 sys + 13.29 cusr  1.47 csys = 14.86 CPU)
Result: FAIL
```

**Comment by MatthewZMD** [comment]
Can someone points me the original fix for this issue? I will try to help as well.

**Comment by jkeenan** [comment]
> Can someone points me the original fix for this issue? I will try to help as well.

@tonycoz @khwilliamson Can you identify the commit with the "original fix"?

I've scanned all files mentioned in this ticket and don't find them mentioned in the *many* commits made on or around Mar 04 2019?

```
git diff -w 635ff1f923d510fc8356bff054b96cbd97d93bf3..aa8a2baaface7f59cac19efdab8e51e08e4d67f6
```
Thank you very much.
Jim Keenan

**Comment by khwilliamson** [comment]
I believe it was a combination of two commits:
commit 8e13243a89cbac8e8110cc8ac0674183cfa30bf7
 Author:     Karl Williamson <khw@cpan.org>
 AuthorDate: Mon Mar 4 12:27:59 2019 -0700
 Commit:     Karl Williamson <khw@cpan.org>
 CommitDate: Mon Mar 4 13:01:37 2019 -0700
 
     Properly handle systems with crippled locales
     
     Some systems fake their locales, so that they pretend to accept a locale
     change, but they either do nothing, making everything the C locale, or
     on some systems there is a a second locale "C-UTF-8" that can be
     switched to.  Configure probes have been added to find such systems, and
     this commit changes to use the results of these probes, so that we don't
     try looking for other locales (any names we came up with would be
     accepted as valid, but don't work, and tests were failing as a result).
     
     Anything running the musl library fits, as does OpenBSD and its kin, as
     they view locales as security risks.  This commit allows us to take out
     some code that was looking for particular OS's.


commit a04a75f20f03aa08ce8118b3b0b3f93eb3e997c5
 Author:     Richard Leach <rich+perl@hyphen-dash-hyphen.info>
 AuthorDate: Sun May 19 20:16:41 2019 +0000
 Commit:     Karl Williamson <khw@cpan.org>
 CommitDate: Fri May 24 17:09:30 2019 -0600
 
*    PATCH: [perl #134117] Close DATA in loc_tools.pl
     
     This prevents unexpected text in lib/warnings.t

**Comment by khwilliamson** [comment]
The first commit avoided some code in problematic systems, but didn't get the underlying issue, which the second should have.

I wonder if Yocto has crippled locale handling, like in musl

**Comment by MatthewZMD** [comment]
Thanks! I'll investigate too

**Comment by MatthewZMD** [comment]
Thanks, I figured that yocto's perl is on 5.30.1, which doesn't have these two commits yet.

I tried to backport a04a75f *only* into the yocto project and rebuild perl on a couple images, `lib/warnings.t` finally passed. (Note that the test passes without needing 8e13243)

I will be submitting the backport patch on the yocto side, unless they decide to upgrade perl to a newer version.

This issue can be closed now. Thank you guys so much for being super helpful!
